File: ChatGPT(예진)/Python (수정 완료)/cpu_monitoring_system.py
import psutil
import threading
import time

# Constants
DEFAULT_THRESHOLD = 90
DEFAULT_DURATION = 10
DEFAULT_MAX_CONTINUOUS_DURATION = 30
DEFAULT_MAX_CONSECUTIVE_THRESHOLD = 95
DEFAULT_INTERVAL = 5

is_monitoring = True

def monitor_keyboard_input():
    global is_monitoring
    while True:
        print("---Press 'p' to stop monitoring---")
        key = input()
        if key == 'p':
            is_monitoring = False
            break
        elif key == 'c':
            print("What action would you like to perform?")
            print("1. Lower process priority")
            print("2. Kill specific process")
            choice = input("Enter your choice (1/2): ")
            if choice == '1':
                process_name = input("Enter the name of the process you want to lower priority for: ")
                try:
                    for proc in psutil.process_iter(['pid', 'name']):
                        if proc.info['name'] == process_name:
                            p = psutil.Process(proc.info['pid'])
                            p.nice(psutil.BELOW_NORMAL_PRIORITY_CLASS)
                            print("Process priority lowered successfully.")
                            break
                except psutil.NoSuchProcess:
                    print("Process not found.")
            elif choice == '2':
                process_name = input("Enter the name of the process you want to kill: ")
                try:
                    for proc in psutil.process_iter(['pid', 'name']):
                        if proc.info['name'] == process_name:
                            p = psutil.Process(proc.info['pid'])
                            p.terminate()
                            print(f"Process '{process_name}' terminated successfully.")
                            break
                except psutil.NoSuchProcess:
                    print("Process not found.")

def monitor_cpu():
    threshold = DEFAULT_THRESHOLD
    consecutive_count = 0
    while is_monitoring:
        cpu_percent = psutil.cpu_percent(interval=DEFAULT_INTERVAL)
        print(f"CPU Usage: {cpu_percent:.2f}%")

        if cpu_percent >= threshold:
            consecutive_count += 1
            if consecutive_count >= DEFAULT_MAX_CONTINUOUS_DURATION / DEFAULT_INTERVAL:
                print("CPU usage exceeded the threshold for too long.")
                print("Taking necessary action...")
                print("The CPU usage has been consistently high. Consider checking for any resource-intensive processes.")
                consecutive_count = 0
        else:
            consecutive_count = 0


def main():
    keyboard_thread = threading.Thread(target=monitor_keyboard_input)
    keyboard_thread.start()

    cpu_thread = threading.Thread(target=monitor_cpu)
    cpu_thread.start()

    keyboard_thread.join()
    cpu_thread.join()

if __name__ == "__main__":
    main()

File: ChatGPT(예진)/Java (수정 완료)/cpu_monitoring_system.java
import java.util.Scanner;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class cpu_monitoring_system {
    // Constants
    private static final int DEFAULT_THRESHOLD = 90;
    private static final int DEFAULT_DURATION = 10;
    private static final int DEFAULT_MAX_CONTINUOUS_DURATION = 30;
    private static final int DEFAULT_MAX_CONSECUTIVE_THRESHOLD = 95;
    private static final int DEFAULT_INTERVAL = 5;

    private static AtomicBoolean isMonitoring = new AtomicBoolean(true);

    public static void main(String[] args) {
        Thread keyboardThread = new Thread(cpu_monitoring_system::monitorKeyboardInput);
        keyboardThread.start();

        Thread cpuThread = new Thread(cpu_monitoring_system::monitorCPU);
        cpuThread.start();

        try {
            keyboardThread.join();
            cpuThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private static void monitorKeyboardInput() {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("---Press 'p' to stop monitoring---");
            String key = scanner.nextLine();
            if (key.equals("p")) {
                isMonitoring.set(false);
                break;
            } else if (key.equals("c")) {
                System.out.println("What action would you like to perform?");
                System.out.println("1. Lower process priority");
                System.out.println("2. Kill specific process");
                String choice = scanner.nextLine();
                if (choice.equals("1")) {
                    System.out.println("Enter the name of the process you want to lower priority for: ");
                    String processName = scanner.nextLine();
                    // Lower process priority logic here
                    System.out.println("Process priority lowered successfully.");
                } else if (choice.equals("2")) {
                    System.out.println("Enter the name of the process you want to kill: ");
                    String processName = scanner.nextLine();
                    // Kill specific process logic here
                    System.out.println("Process '" + processName + "' terminated successfully.");
                }
            }
        }
    }

    private static void monitorCPU() {
        int threshold = DEFAULT_THRESHOLD;
        AtomicInteger consecutiveCount = new AtomicInteger(0);
        while (isMonitoring.get()) {
            double cpuPercent = getCpuUsage();
            System.out.println("CPU Usage: " + String.format("%.2f", cpuPercent) + "%");

            if (cpuPercent >= threshold) {
                consecutiveCount.incrementAndGet();
                if (consecutiveCount.get() >= DEFAULT_MAX_CONTINUOUS_DURATION / DEFAULT_INTERVAL) {
                    System.out.println("CPU usage exceeded the threshold for too long.");
                    System.out.println("Taking necessary action...");
                    System.out.println("The CPU usage has been consistently high. Consider checking for any resource-intensive processes.");
                    consecutiveCount.set(0);
                }
            } else {
                consecutiveCount.set(0);
            }
            try {
                Thread.sleep(DEFAULT_INTERVAL * 1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    private static double getCpuUsage() {
        // Logic to get CPU usage
        return 0.0;
    }
}
